sql/core/src/main/scala/org/apache/spark/sql/streaming/StreamingRepository.scala

```diff
@@ -0,0 +1,177 @@
+ * may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License. See accompanying
+ * LICENSE file.
+ */
+
+package org.apache.spark.sql.streaming
+
+import java.text.SimpleDateFormat
+import java.util.{Date, TimeZone, UUID}
+
+import scala.collection.mutable.HashMap
+
+import org.apache.commons.collections.buffer.CircularFifoBuffer
+
+import org.apache.spark.ui.UIUtils
+
+class StreamingRepository {
+
+  val allQueries = HashMap.empty[UUID, StreamingQueryStatistics]
+
+}
+
+object StreamingRepository {
+
+  private val _instance: StreamingRepository = new StreamingRepository
+
+  def getInstance: StreamingRepository = _instance
+}
+
+
+class StreamingQueryStatistics (
+    qId: UUID,
+    qName: String,
+    runId: UUID,
+    startTime: Long,
+    trigger: Trigger = ProcessingTime(0L)) {
+
+  private val MAX_SAMPLE_SIZE = 100
+
+  private val simpleDateFormat = new SimpleDateFormat("dd-MMM-YYYY hh:mm:ss")
+  private val timestampFormat = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'") // ISO8601
+  timestampFormat.setTimeZone(TimeZone.getTimeZone("UTC"))
+
+  val queryUUID: UUID = qId
+  val queryName: String = qName
+  val queryStartTime: Long = startTime
+  val queryStartTimeText: String = simpleDateFormat.format(startTime)
+  var queryUptime: Long = 0L
+  var queryUptimeText: String = ""
+
+  var runUUID: UUID = runId
+  val trendEventsInterval: Long = trigger.asInstanceOf[ProcessingTime].intervalMs
+
+  var isActive: Boolean = true
+
+  var currentBatchId: Long = -1L
+
+  var sources = Array.empty[SourceProgress]
+  var sink: SinkProgress = null
+
+  var totalInputRows: Long = 0L
+  var avgInputRowsPerSec: Double = 0.0
+  var avgProcessedRowsPerSec: Double = 0.0
+  var totalProcessingTime: Long = 0L
+  var avgProcessingTime: Double = 0.0
+  var numBatchesProcessed: Long = 0L
+
+  val timeLine = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+  val numInputRowsTrend = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+  val inputRowsPerSecondTrend = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+  val processedRowsPerSecondTrend = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+  val processingTimeTrend = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+  val batchIds = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+
+  var currStateOpNumRowsTotal = 0L
+  var currStateOpNumRowsUpdated = 0L
+  val stateOpNumRowsTotalTrend = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+  val stateOpNumRowsUpdatedTrend = new CircularFifoBuffer(MAX_SAMPLE_SIZE)
+
+  def updateQueryStatistics(event: StreamingQueryListener.QueryProgressEvent): Unit = {
+    val progress = event.progress
+
+    if(this.currentBatchId < progress.batchId){
```

Comment:
Please see if we need to ignore or overwrite the existing stats in case of event for older batch id.

Reply:
Although I have implemented code changes to handle restart query case, I haven't added those changes yet. Some more discussion is needed.
So I suggest not to include this in current release. We can take it as an enhancement in next release.

Reply:
Ok. Lets file that as a separate task and checkin.